### DISCUSSION TITLE
OpenTherm implementation

### DISCUSSION THREAD

**Discussion by robertugo2** [mariusz-ostoja-swierczynski/tech-controllers#129] [Ideas]
Hi,

Recently I have setup L-12 controller. This controller supports OpenTherm protocol. Currently I'm trying to implement support for that.
Example of tile daya:
```
{
	'id': 8009, 
	'parentId': 0, 
	'type': 252, 
	'menuId': 1053, 
	'orderId': None, 
	'visibility': True, 
	'params': {
		'description': 'OpenTherm', 
		'currentTemp': 283, 
		'setCurrentTemp': 296, 
		'currentTempDHW': 457, 
		'setTempDHW': 470, 
		'modulationPercentage': 0, 
		'flags': {
			'communication': 0, 
			'heatingCurve': 1, 
			'activeHeating': 1, 
			'activeDHW': 0
		}, 
		'alarmCode': 0}
}
```
I'm thinking how to implement support for this tile to be compatible with current codding style. Also I don't know how OpenTherm works for other furnaces, that is if all data listed above are available.

My ideas are:
* Create some super class for temperature sensors and then child classes for individual measurements. That includes already existing temperature sensors.
* Create individual classes for `setCurrentTemp`, `setCurrentTemp` and so on, that inherits from `TileSensor` and `SensorEntity`

In `async_setup_entry` function initialization will look like as follows:
```
        if tile[CONF_TYPE] == TYPE_OPEN_THERM:
            params = tile[CONF_PARAMS]
            if params.get(PARAM_OPENTHERM_SET_CURRENT_TEMP) is not None:
                entities.append(TileOpenThermSetTemperatureSensor(tile, coordinator, config_entry))
            if params.get(PARAM_OPENTHERM_CURRENT_TEMP) is not None:
                entities.append(TileOpenThermCurrenltTemperatureSensor(tile, coordinator, config_entry))
            if params.get(PARAM_OPENTHERM_DHW_SET_CURRENT_TEMP) is not None:
                entities.append(TileOpenThermDHWSetTemperatureSensor(tile, coordinator, config_entry))
            if params.get(PARAM_OPENTHERM_DHW_CURRENT_TEMP) is not None:
                entities.append(TileOpenThermDHWCurrenltTemperatureSensor(tile, coordinator, config_entry))
```
I don't know if all parameters exists on all OpenTherm devices so there is a check, if given parameter exists.

**Comment by daroga0002** [comment]
Opentherm is lower layer protocol as per https://ihormelnyk.com/Content/Pages/opentherm_library/Opentherm%20Protocol%20v2-2.pdf?srsltid=AfmBOoq_TxCDE8DIcIBfa7GhisoRcE5f3kRodfNIALzJAYdTkX1LJZIQ


so what you really what probably to achieve is adding support to devices which are represented as OpenTherm devices (but this doesnt mean you will implement Opentherm protocol).



Good starting point will be probably to put  here full API outputs which contains example outputs from tech API

**Comment by robertugo2** [comment]
Thanks for your reply. Below you can see truncated response for module data:
```
{
   "zones":{
	   ...
   },
   "tiles":[
       ...
      {
         "id":8009,
         "parentId":0,
         "type":252,
         "menuId":1053,
         "orderId":"None",
         "visibility":true,
         "params":{
            "description":"OpenTherm",
            "currentTemp":314,
            "setCurrentTemp":323,
            "currentTempDHW":464,
            "setTempDHW":470,
            "modulationPercentage":0,
            "flags":{
               "communication":0,
               "heatingCurve":1,
               "activeHeating":1,
               "activeDHW":0
            },
            "alarmCode":0
         }
      }
   ],
   "tilesOrder":"None",
   "tilesLastUpdate":"2024-11-04 22:01:33.614189+01"
}
```
For now (I haven't tested that yet) I have defined this class:
```
class OpenThermSensor(TileSensor, SensorEntity):
    """Representation of a OpenTherm device."""

    def __init__(self, device, coordinator, config_entry) -> None:
        """Initialize the sensor."""
        TileSensor.__init__(self, device, coordinator, config_entry)
        self.native_unit_of_measurement = UnitOfTemperature.CELSIUS
        self.device_class = SensorDeviceClass.TEMPERATURE
        self.state_class = SensorStateClass.MEASUREMENT
        self._attr_icon = assets.get_icon_by_type(device[CONF_TYPE])
        self._name = (
            self._config_entry.title + " "
            if self._config_entry.data[INCLUDE_HUB_IN_NAME]
            else ""
        ) + assets.get_text_by_type(device[CONF_TYPE])

        self.attrs: dict[str, Any] = {}

    @property
    def unique_id(self) -> str:
        """Return a unique ID."""
        return f"{self._unique_id}_tile_opentherm"

    @property
    def name(self) -> str | UndefinedType | None:
        """Return the name of the device."""
        return f"{self._name}"

    def get_state(self, device) -> Any:
        """Get the state of the device."""
        return device[CONF_PARAMS][OPENTHERM_CURRENT_TEMP]/10

    @property
    def extra_state_attributes(self) -> dict[str, Any]:
        """Return the state attributes."""
        attributes = {}
        attributes.update(self.attrs)
        return attributes

    def update_properties(self, device):
        """Update the properties of the device based on the provided device information.

        Args:
        device: dict, the device information containing description, zone, setTemperature, and currentTemperature

        Returns:
        None

        """
        self._state = self.get_state(device)
        self.attrs[OPENTHERM_CURRENT_TEMP] = device[CONF_PARAMS][OPENTHERM_CURRENT_TEMP] / 10
        self.attrs[OPENTHERM_CURRENT_TEMP_DHW] = device[CONF_PARAMS][OPENTHERM_CURRENT_TEMP_DHW] / 10
        self.attrs[OPENTHERM_SET_CURRENT_TEMP] = device[CONF_PARAMS][OPENTHERM_SET_CURRENT_TEMP] / 10
        self.attrs[OPENTHERM_SET_TEMP_DHW] = device[CONF_PARAMS][OPENTHERM_SET_TEMP_DHW] / 10
```
Do you think that what I made have any sense?

**Reply by daroga0002** [reply]
on first look it make sense, but I propose fork repo, create branch, make changes and then raise a PR request for review

**Reply by robertugo2** [reply]
Ok, thanks. But currently I'm having problems with how to define and based on what I should implement it.
In UI open therm is displayed like external temperature:
![image](https://github.com/user-attachments/assets/573b6893-796a-45f5-915a-cb2bd9de57b5)
So I assumed, that it will be implemented like `TileTemperatureSensor`, but I think my assumption was wrong as external temperature sensor is implemented using "ext_temperature_entity" key from "strings.json". This key is used by:
* TechOutsideTempTile
* ZoneOutsideTempTile
First one seems to be not used anywhere as I can't find any initialization procedure for this class. ZoneOutsideTempTile seems to be not used for general type of device and seems to be linked to particular zone.

So, based on what I should develop module for OpenTherm? There will be 4 values as stated previously.

**Reply by daroga0002** [reply]
External temperature sensor is arleady implemented here https://github.com/mariusz-ostoja-swierczynski/tech-controllers/blob/910f035a6acd79e57b2a46bba82775c3f19701c9/custom_components/tech/sensor.py#L1265

I think code which you pasted here https://github.com/mariusz-ostoja-swierczynski/tech-controllers/discussions/129#discussioncomment-11147415 will be good starting point as I dont know what can change in this entity tile
